modules/basics/src/main/java/com/opengamma/basics/currency/MultiCurrencyAmount.java

```diff
@@ -145,12 +132,47 @@ public static MultiCurrencyAmount of(Map<Currency, Double> map) {
    */
   public static MultiCurrencyAmount total(Iterable<CurrencyAmount> amounts) {
     ArgChecker.notNull(amounts, "amounts");
-    Map<Currency, CurrencyAmount> map = new HashMap<Currency, CurrencyAmount>();
-    for (CurrencyAmount currencyAmount : amounts) {
-      ArgChecker.notNull(currencyAmount, "amount");
-      map.merge(currencyAmount.getCurrency(), currencyAmount, CurrencyAmount::plus);
-    }
-    return new MultiCurrencyAmount(ImmutableSortedSet.copyOf(map.values()));
+    return Guavate.stream(amounts).collect(collector());
+  }
+
+  //-------------------------------------------------------------------------
+  /**
+   * Returns a collector that can be used to create a multi-currency amount from a stream of amounts.
+   * <p>
+   * If the input contains the same currency more than once, the amounts are added together.
+   * For example, an input of (EUR 100, EUR 200, CAD 100) would result in (EUR 300, CAD 100).
+   *
+   * @return the collector
+   */
+  public static Collector<CurrencyAmount, ?, MultiCurrencyAmount> collector() {
+    return Collector.<CurrencyAmount, Map<Currency, CurrencyAmount>, MultiCurrencyAmount>of(
+            // accumulate into a map
+            HashMap::new,
+            // merge two CurrencyAmounts if same currency
+            (map, ca) -> map.merge(ArgChecker.notNull(ca, "amount").getCurrency(), ca, CurrencyAmount::plus),
+            // combine two maps
+            (map1, map2) -> {
+              map1.putAll(map2);
+              return map1;
+            },
+            // convert to MultiCurrencyAmount
+            (map) -> new MultiCurrencyAmount(ImmutableSortedSet.copyOf(map.values())),
+            UNORDERED);
+  }
+
+  /**
+   * Returns a collector that can be used to create a multi-currency amount
+   * from a stream of amounts where each amount has a different currency.
+   * <p>
+   * Each amount in the stream must have a different currency.
+   *
+   * @return the collector
+   */
+  private static Collector<CurrencyAmount, ?, MultiCurrencyAmount> collectorInternal() {
+    // this method must not be exposed publicly as misuse creates an instance with invalid state
```

Comment:
Added a comment to explain - when used internally it avoids unnecessary work thus will result in better performance